Automattic/wp-calypso
# Jetpack Cloud: improve our approach to fixing multiple threats

#### Changes proposed in this Pull Request

Things this PR fixes:
* Currently, when a user tries to fix all threats, a notice is displayed for every threat. Now this PR changes that and only one notice is displayed.
* Because fixing/ignoring a threat happens in the background and it can take any amount of time, sometimes a threat in the UI stays forever as disabled (in progress) until the user moves to another page or refresh the tab. This is how the process works now:

```
            t1                t2                       t3                t4
request fix/ignore ---> operation succeeded ---> request scan state ---> ????
```

It can happen that even in t4, the threat has not been fixed/ignored yet. If this happens, the UI stays the same until the users do something that triggers a new request for the state of the Scan. This PR fix that by pooling until we know all threats have been fixed.

#### Implementation notes

The new pooling mechanism works only if the following conditions are true:

* The action includes `pooling: true`
* The scan state is `idle`
* There are threats that are being fixed/ignored (`fixerStatus === 'in_progress'`)

This means, the app will keep asking `/scan` every `POOL_EVERY_MILLISECONDS` until all threats are fixed/ignored.

#### Testing instructions

* Go to http://jetpack.cloud.localhost:3000/
* Select a site that has Scan and at least two threats
* Go to Scan -> Scanner
* Click the `[Fix all]` button and confirm the action
* Verify only one notice is displayed
* Verify request are sent to `/scan` until all threats are fixed (no threat has the `fixer_status` property)
* Verify all threats are remove from the screen and you are back to the Scan is okay UI

Fixes 1151678672052943-as-1175373711120913

#### Demo

![Kapture 2020-05-13 at 15 53 17](https://user-images.githubusercontent.com/3418513/81852854-0b608600-9532-11ea-91a5-488860e15d47.gif)


**If the state of the Scan looks like this, the app should keep pooling until all `fixer_status` properties disappear from the threats.**
![image](https://user-images.githubusercontent.com/3418513/81852911-23d0a080-9532-11ea-9adb-424c1e7936a3.png)
